cudax/include/cuda/experimental/__container/vector.cuh

```diff
@@ -0,0 +1,1585 @@
+    }
+    else if (__pos != __size_)
+    { // need to create space in the middle
+      _CCCL_IF_CONSTEXPR (__is_host_only)
+      {
+        _CUDA_VSTD::move_backward(__middle, __end, __end + __count);
+      }
+      else
+      { // on device we need temporary storage
+        __buffer __temp_buf{__borrow_resource(), static_cast<size_type>(__size_ - __pos)};
+        this->__copy_same(__middle, __end, __temp_buf.begin());
+        this->__copy_same(__temp_buf.begin(), __temp_buf.end(), __middle + __count);
+      }
+    }
+  }
+
+  //! @brief Rotates elements from `[__first, __middle)` after `[__middle, __last)`.
+  //! @param __first Pointer to the start of the first segment.
+  //! @param __middle Pointer to the end of the first segment.
+  //! @param __last Pointer to the end of the second segment.
+  void __rotate(pointer __first, pointer __middle, pointer __last)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      _CUDA_VSTD::rotate(__first, __middle, __last);
+    }
+    else
+    { // on device we need temp storage
+      const auto __size = static_cast<size_type>(__last - __first);
+      __buffer __temp_buf{__borrow_resource(), __size};
+      pointer __tail = this->__copy_same(__middle, __last, __temp_buf.begin());
+      this->__copy_same(__first, __middle, __tail);
+      this->__copy_same(__temp_buf.begin(), __temp_buf.end(), __first);
+    }
+  }
+
+  //! @brief Value-initializes elements in the range `[__first, __first + __count)`.
+  //! @param __first Pointer to the first element to be initialized.
+  //! @param __count The number of elements to be initialized.
+  static void __value_initialize_n(pointer __first, size_type __count)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      _CUDA_VSTD::fill_n(__first, __count, _Tp());
+    }
+    else
+    {
+      thrust::fill_n(thrust::device, __first, __count, _Tp());
+    }
+  }
+
+  //! @brief Copy-constructs elements in the range `[__first, __first + __count)`.
+  //! @param __first Pointer to the first element to be initialized.
+  //! @param __count The number of elements to be initialized.
+  static void __fill_n(pointer __first, size_type __count, const _Tp& __value)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      _CUDA_VSTD::fill_n(__first, __count, __value);
+    }
+    else
+    {
+      thrust::fill_n(thrust::device, __first, __count, __value);
+    }
+  }
+
+  //! @brief Removes elements within the sequence `[__first, __last)` that are equal to \p __value.
+  //! @param __first Pointer to the start of the sequence.
+  //! @param __last Pointer to the end of the sequence.
+  //! @param __value The element to be removed.
+  static pointer __remove_value(pointer __first, pointer __last, const _Tp& __value)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      return _CUDA_VSTD::remove(__first, __last, __value);
+    }
+    else
+    {
+      return thrust::remove(thrust::device, __first, __last, __value);
```

Comment:
If I understand thrust will use NULL stream by default, should we make change it to use some internal stream? Otherwise you might run into some very bad synchronization issues